feat(ec2): flow logs from TransitGateway and TransitGatewayAttachment

badmintoncryer opened the pull request:
I have enabled the configuration of flow logs for TransitGateway and TransitGatewayAttachment. 

Create flow logs from TransitGateway:
```ts
declare const tgw: ec2.CfnTransitGateway;

new ec2.FlowLog(this, 'TransitGatewayFlowLog', {
  resourceType: ec2.FlowLogResourceType.fromTransitGatewayId(tgw.ref)
})
```

Create flowlogs from TransitGatewayAttachment:
```ts
declare const tgwAttachment: ec2.CfnTransitGatewayAttachment;

new ec2.FlowLog(this, 'TransitGatewayAttachmentFlowLog', {
  resourceType: ec2.FlowLogResourceType.fromTransitGatewayAttachmentId(tgwAttachment.ref)
})
```

Since `trafficType` [cannot be set for flow logs related to TransitGateway resources](https://docs.aws.amazon.com/vpc/latest/tgw/working-with-flow-logs.html#flow-logs-api-cli), I have also added error handling for this.

```ts
if (props.resourceType.resourceType === 'TransitGateway' || props.resourceType.resourceType === 'TransitGatewayAttachment') {
      if (props.trafficType) {
        throw new Error('trafficType is not supported for Transit Gateway and Transit Gateway Attachment');
      }
      ...
    }
```

Closes #27222.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (changes requested):
The pull request linter has failed. See the aws-cdk-automation comment below for failure reasons. If you believe this pull request should receive an exemption, please comment and provide a justification.


A comment requesting an exemption should contain the text `Exemption Request`. Additionally, if clarification is needed add `Clarification Request` to a comment.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to a README file.
	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to a README file.
	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 1dbb2b8e02783b75e55c7e37b58624c1dde90af9
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0e3bb639-e8c9-4579-b359-493a7a60c23d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: d02831881909881448858139ee65fea754d73ff6
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4dbab24b-16ca-4c91-924a-551170e8de23%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a9e17205a596450cc126e5e35b46b085d3bd2c15
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c469884d-ac5a-4723-9c9b-5967790d8259%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: fe588a40fbb36e56a23dbc524219ff8b345f67e4
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1d197270-775c-46a9-9ed8-0474213b2a15%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: b779cb7ca5f5c564b606f688aa3dd07b86840260
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=65cb7854-9b9e-4bca-8906-c947b41257f8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 12c8bcbd1600f069f7ee9d1d9c585151e8d84d86
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f6a7c0db-9588-4aef-97b9-36132808a810%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

go-to-k reviewed (changes requested):
Validation of `maxAggregationInterval` might be a good idea in this PR.
Also, it might be good to comment the description for the validation in the doc of the `maxAggregationInterval` parameter.

> --max-aggregation-interval has a default value of 60, and is the only accepted value for transit gateway resource types. An error is returned if you try to pass any other value. This limit applies only to transit gateway resource types.

https://docs.aws.amazon.com/vpc/latest/tgw/working-with-flow-logs.html

> This parameter must be 60 seconds for transit gateway resource types.

https://docs.aws.amazon.com/ja_jp/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-flowlog.html#cfn-ec2-flowlog-maxaggregationinterval

go-to-k commented:
And this PR title `feat(ec2): flow logs from TransitGateway and TransitGatewayAttachment ` would be better.

go-to-k commented on packages/aws-cdk-lib/aws-ec2/lib/vpc-flow-logs.ts:
```diff
@@ -800,6 +823,17 @@ export class FlowLog extends FlowLogBase {
     }
 
+    // In Transit Gateway and Transit Gateway Attachment, `trafficType` is not supported.
+    let trafficType: FlowLogTrafficType | undefined = FlowLogTrafficType.ALL;
+    if (props.resourceType.resourceType === 'TransitGateway' || props.resourceType.resourceType === 'TransitGatewayAttachment') {
+      if (props.trafficType) {
+        throw new Error('trafficType is not supported for Transit Gateway and Transit Gateway Attachment');
+      }
+      trafficType = undefined;
+    } else if (props.trafficType) {
+      trafficType = props.trafficType;
+    }
```
Refactoring may be good.

```suggestion
    let trafficType: FlowLogTrafficType | undefined = props.trafficType ?? FlowLogTrafficType.ALL;
    if (props.resourceType.resourceType === 'TransitGateway' || props.resourceType.resourceType === 'TransitGatewayAttachment') {
      if (props.trafficType) {
        throw new Error('trafficType is not supported for Transit Gateway and Transit Gateway Attachment');
      }
      trafficType = undefined;
    }
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 3aa77d9b1686dc39dc16b1c321f299924ac0973b
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a642b49d-a58a-432c-8002-7cbfc50e5f7a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ab41d4e1e530cfdce87fc7ad3f0ff38741836ee9
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=852d94bd-b132-4a45-a7e3-d250b6bd8c48%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: d3e85e6239dc4687007eebb279c83bda941d8322
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=77925236-47bd-4aaa-b9b9-19f696e5e645%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

badmintoncryer commented:
@go-to-k 
Thank you for your review! I have incorporated your suggested modifications.

I've implemented validation of `maxAggregationInterval`.
The default value for this parameter is 10 minutes, which should result in an error when creating a Transit Gateway flow log. However, in practice, CFN allows deployment when `maxAggregationInterval` is set to undefined. It's likely that, internally, the default value is adjusted to 1 minute for Transit Gateway targets.

Therefore, when the target is TransitGateway, we allow `maxAggregationInterval` to be either `ONE_MINUTE` or `undefined`, and only consider it an error if `TEN_MINUTES` is explicitly specified."

go-to-k commented on packages/aws-cdk-lib/aws-ec2/lib/vpc-flow-logs.ts:
```diff
@@ -688,6 +711,11 @@ export interface FlowLogOptions {
    * The maximum interval of time during which a flow of packets is captured
    * and aggregated into a flow log record.
    *
+   * When creating flow logs for a Transit Gateway or Transit Gateway Attachment,
+   * it is not possible to specify TEN_MINUTES.
+   *
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-flowlog.html#cfn-ec2-flowlog-maxaggregationinterval
+   *
    * @default FlowLogMaxAggregationInterval.TEN_MINUTES
```
I changed the description and the default.

```suggestion
   * When creating flow logs for a Transit Gateway or Transit Gateway Attachment,
   * this property must be ONE_MINUTES.
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-flowlog.html#cfn-ec2-flowlog-maxaggregationinterval
   *
   * @default - FlowLogMaxAggregationInterval.ONE_MINUTES if Transit Gateway or
   * Transit Gateway Attachment, otherwise FlowLogMaxAggregationInterval.TEN_MINUTES.
```

go-to-k commented on packages/aws-cdk-lib/aws-ec2/test/vpc-flow-logs.test.ts:
```diff
@@ -542,6 +542,88 @@ describe('vpc flow logs', () => {
+    });
+  });
+
+  test('create with transit gateway id and specifiy traffic type', () => {
```
```suggestion
  test('throw if transit gateway with traffic type', () => {
```

go-to-k commented on packages/aws-cdk-lib/aws-ec2/test/vpc-flow-logs.test.ts:
```diff
@@ -542,6 +542,88 @@ describe('vpc flow logs', () => {
     });
   });
+
+  test('create with transit gateway id`', () => {
```
```suggestion
  test('create from transit gateway`', () => {
```

go-to-k commented on packages/aws-cdk-lib/aws-ec2/test/vpc-flow-logs.test.ts:
```diff
@@ -542,6 +542,88 @@ describe('vpc flow logs', () => {
+    }).toThrow(/trafficType is not supported for Transit Gateway and Transit Gateway Attachment/);
+  });
+
+  test('create with transit gateway attachment id', () => {
```
```suggestion
  test('create from transit gateway attachment', () => {
```

go-to-k commented on packages/aws-cdk-lib/aws-ec2/test/vpc-flow-logs.test.ts:
```diff
@@ -542,6 +542,88 @@ describe('vpc flow logs', () => {
+    });
+  });
+
+  test('create with transit gateway attachment id and specifiy traffic type', () => {
```
```suggestion
  test('throw if transit gateway attachment with traffic type', () => {
```

go-to-k commented on packages/aws-cdk-lib/aws-ec2/test/vpc-flow-logs.test.ts:
```diff
@@ -542,6 +542,88 @@ describe('vpc flow logs', () => {
+    }).toThrow(/trafficType is not supported for Transit Gateway and Transit Gateway Attachment/);
+  });
+
+  test('create with transit gateway id and specify maxAggregationInterval.TEN_MINUTES', () => {
```
```suggestion
  test('throw if transit gateway with value other than maxAggregationInterval.ONE_MINUTES', () => {
```

go-to-k commented on packages/aws-cdk-lib/aws-ec2/test/vpc-flow-logs.test.ts:
```diff
@@ -542,6 +542,88 @@ describe('vpc flow logs', () => {
+    }).toThrow(/maxAggregationInterval cannot be set to TEN_MINUTES for Transit Gateway resources/);
+  });
+
+  test('create with transit gateway attachment id and specify maxAggregationInterval.TEN_MINUTES', () => {
```
```suggestion
  test('throw if transit gateway attachment with value other than maxAggregationInterval.ONE_MINUTES', () => {
```

go-to-k reviewed (changes requested):
Thanks, I've made a few more adjustment comments.

go-to-k commented on packages/aws-cdk-lib/aws-ec2/lib/vpc-flow-logs.ts:
```diff
@@ -775,6 +803,13 @@ export class FlowLog extends FlowLogBase {
   constructor(scope: Construct, id: string, props: FlowLogProps) {
     super(scope, id);
 
+    if (
+      ['TransitGateway', 'TransitGatewayAttachment'].some((type) => type === props.resourceType.resourceType)
+      && props.maxAggregationInterval === FlowLogMaxAggregationInterval.TEN_MINUTES
+    ) {
+      throw new Error('maxAggregationInterval cannot be set to TEN_MINUTES for Transit Gateway resources');
+    }
+
```
https://github.com/aws/aws-cdk/pull/28605#discussion_r1448588949

badmintoncryer commented on packages/aws-cdk-lib/aws-ec2/lib/vpc-flow-logs.ts:
```diff
@@ -800,6 +835,15 @@ export class FlowLog extends FlowLogBase {
       }).join(' ');
     }
 
+    // In Transit Gateway and Transit Gateway Attachment, `trafficType` is not supported.
+    let trafficType: FlowLogTrafficType | undefined = props.trafficType ?? FlowLogTrafficType.ALL;
+    if (props.resourceType.resourceType === 'TransitGateway' || props.resourceType.resourceType === 'TransitGatewayAttachment') {
+      if (props.trafficType) {
+        throw new Error('trafficType is not supported for Transit Gateway and Transit Gateway Attachment');
+      }
+      trafficType = undefined;
+    }
+
```
Thank you for your suggestion. 

I fully agree with moving the error handling to this location. 
(In fact, I'm embarrassed about the initial implementation where I wrote it in a very redundant place😢)

I have a question. In the implementation example you proposed, it seems that there would be an error when `maxAggregationInterval` is `undefined`. Is that acceptable? 
According to the documentation, the default value of `maxAggregationInterval` is 10 minutes, and for TransitGateway, only 1 minute can be set. However, when deploying the CFN with `maxAggregationInterval` left as undefined, the deployment proceeds normally. 

Therefore, I thought it would be acceptable to allow undefined as well. (I'm concerned about causing new errors for users who have traditionally been creating resources with maxAggregationInterval set to undefined.)

go-to-k replied:
> In the implementation example you proposed, it seems that there would be an error when maxAggregationInterval is undefined

Really? I did not envision that happening with my proposed code. I don't think the following code I just suggested would cause the error, but did it happen elsewhere?

(I think the error do not occur by `props.maxAggregationInterval && `.)

```ts
    let trafficType: FlowLogTrafficType | undefined = props.trafficType ?? FlowLogTrafficType.ALL;
    if (props.resourceType.resourceType === 'TransitGateway' || props.resourceType.resourceType === 'TransitGatewayAttachment') {
      if (props.trafficType) {
        throw new Error('trafficType is not supported for Transit Gateway and Transit Gateway Attachment');
      }
      if (props.maxAggregationInterval && props.maxAggregationInterval !== FlowLogMaxAggregationInterval.ONE_MINUTE) {
        throw new Error('maxAggregationInterval must be set to ONE_MINUTE for Transit Gateway and Transit Gateway Attachment');
      }
      trafficType = undefined;
    }
```

badmintoncryer replied:
@go-to-k 
oh.. I'm sorry. I overlooked `props.maxAggregationInterval`.

go-to-k commented:
> In the implementation example you proposed, it seems that there would be an error when maxAggregationInterval is undefined

Really? I did not envision that happening with my proposed code. I don't think the following code I just suggested would cause the error, but did it happen elsewhere?

(I think the error do not occur by `props.maxAggregationInterval && `.)

```ts
    let trafficType: FlowLogTrafficType | undefined = props.trafficType ?? FlowLogTrafficType.ALL;
    if (props.resourceType.resourceType === 'TransitGateway' || props.resourceType.resourceType === 'TransitGatewayAttachment') {
      if (props.trafficType) {
        throw new Error('trafficType is not supported for Transit Gateway and Transit Gateway Attachment');
      }
      if (props.maxAggregationInterval && props.maxAggregationInterval !== FlowLogMaxAggregationInterval.ONE_MINUTE) {
        throw new Error('maxAggregationInterval must be set to ONE_MINUTE for Transit Gateway and Transit Gateway Attachment');
      }
      trafficType = undefined;
    }
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a83658f43b6533e1c9f9b172e0b683b4453351a6
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f792495f-7425-47ff-a5ac-9116ff32d054%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 49a9f89d9f68f31567e7f16a4fb69ce6eccccfa8
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cc3091a0-8af4-4ab8-b854-20164c5ce9a9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 27f4c10bf19d7b33721137e36f4ac380b7c683e2
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6a75789d-5d70-450f-ab55-dcc28df3cda1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 50d230c7cb553cdc8cbe9311b18a2721cbb5dd4a
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=40d9de7c-1bb4-4d48-ad92-8ab2c76a1342%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 65f6d2b58e1affb3fb2b9019051dfdb480d8f10c
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f7e4335b-4561-4f94-ada4-774841e586ef%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 515a11ad4326db9d27fc5954978c91c5ccf74300
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f2df84a2-c493-46b3-bbb7-0732e71eeab0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 4765687cc18db43b0add3b862ae291ff314b0e83
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d1543d74-e53c-4eb3-925a-986a54288a07%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: fd3e10cc07ec76050ad9de2252d188b048676e22
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=436b8f6b-ae64-441b-8af0-d24e8829c7d7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 1ec14a76eb5618a2f46815f47589ec27caf40a56
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fc454a63-62b1-4560-9019-c32323e149eb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

go-to-k reviewed (approved):
Looks good.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: d9b1f4eb29cac7ec9b346e943ef712562525975f
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a1222481-f50f-4357-a256-cde358b2008b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a30109b3ab2ea9f661cf5f404d1ed15d4d7d3012
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f5369b62-8a63-429e-a1d2-b770ee74bcd7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 4f53223a81e90d48c914fa790cdf147355f7c6d5
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e1d9172c-e4d1-49d5-b71c-3028e72d6e75%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

paulhcsun commented on packages/aws-cdk-lib/aws-ec2/lib/vpc-flow-logs.ts:
```diff
@@ -659,6 +679,9 @@ export class LogFormat {
 export interface FlowLogOptions {
   /**
    * The type of traffic to log. You can log traffic that the resource accepts or rejects, or all traffic.
+   * When the target is either TransitGateway or TransitGatewayAttachment, setting the traffic type is not possible.
```
```suggestion
   * When the target is either `TransitGateway` or `TransitGatewayAttachment`, setting the traffic type is not possible.
```

paulhcsun commented on packages/aws-cdk-lib/aws-ec2/lib/vpc-flow-logs.ts:
```diff
@@ -688,7 +711,13 @@ export interface FlowLogOptions {
+   *
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-flowlog.html#cfn-ec2-flowlog-maxaggregationinterval
+   *
+   * @default - FlowLogMaxAggregationInterval.ONE_MINUTES if Transit Gateway or
```
Could we update this to say something like `FlowLogMaxAggregationInterval.ONE_MINUTES if creating flow logs for Transit Gateway or....` or `FlowLogMaxAggregationInterval.ONE_MINUTES if target is Transit Gateway or...`? Just saying `if Transit Gateway or Transit Gateway Attachment` doesn't feel very clear.

paulhcsun commented on packages/aws-cdk-lib/aws-ec2/lib/vpc-flow-logs.ts:
```diff
@@ -800,6 +829,17 @@ export class FlowLog extends FlowLogBase {
     }
 
+    let trafficType: FlowLogTrafficType | undefined = props.trafficType ?? FlowLogTrafficType.ALL;
+    if (props.resourceType.resourceType === 'TransitGateway' || props.resourceType.resourceType === 'TransitGatewayAttachment') {
```
Can you create constants or enums for the `resourceType`?

badmintoncryer replied:
@paulhcsun 
I agree with creating an enum! 
However, in the existing code, a `resourceType` defined as a string is already being used. Along with defining the enum, is it okay to also modify the `FlowLogResourceType`'s `ResourceType` variable and the methods `fromSubnet()`, `fromVpc()`, and `fromNetworkInterfaceId()`?

```ts
const enum ResourceType {
  TRANSIT_GATEWAY = 'transitGateway',
  ...
}

export abstract class FlowLogResourceType {
  public static fromSubnet(subnet: ISubnet): FlowLogResourceType {
    return {
      resourceType: ResourceType.SUBNET,
      resourceId: subnet.subnetId,
    };
  }
...
  /**
   * The type of resource to attach a flow log to.
   */
  public abstract resourceType: ResourceType;
...
}

```

badmintoncryer replied:
@paulhcsun I see! I may be create PR for that later.

kaizencc replied:
@badmintoncryer I don't think doing so would be safe so I would hold off on that for now

badmintoncryer replied:
@kaizencc Understood. Thank you for the additional information.

paulhcsun commented on packages/aws-cdk-lib/aws-ec2/test/vpc-flow-logs.test.ts:
```diff
@@ -542,6 +542,88 @@ describe('vpc flow logs', () => {
+    }).toThrow(/maxAggregationInterval must be set to ONE_MINUTE for Transit Gateway and Transit Gateway Attachment/);
+  });
+
+  test('throw if transit gateway attachment with value other than maxAggregationInterval.ONE_MINUTES', () => {
```
```suggestion
  test('throw if transit gateway attachment with maxAggregationInterval value other than maxAggregationInterval.ONE_MINUTES', () => {
```

paulhcsun reviewed (changes requested):
Nice work @badmintoncryer, looks good overall, just a few minor comments to address.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 38d1e893f6a5a5ac1c69d70a2087c981aa24f708
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=caa1feb1-828e-4dd2-b270-2c0d1979759f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 706b527e17785c62286e35de600491ee6aad74c4
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2e1bfa73-bcec-4b73-8acc-549d3c70e57e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 9f9b1a2dcd2912f9bffa177c205688bb5ac7a7a6
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=258eb586-8cae-414a-a2cc-83faf774fd54%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 6147412b8307bee1fc22d3c5a46e41f53a70c7c4
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8f0b05f9-da81-441a-ab2c-f363bee56755%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 95840743b5d70c8e98a5311a8716f1853bbe8dfe
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c1efbd66-5324-45f6-b338-9feea0f080db%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 3a14a7e9adac5c377842708ea075a651993c224e
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=eef4d1f4-128f-491d-91c1-61e92f79356a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

paulhcsun reviewed (approved):
Great work @badmintoncryer! 

And thanks for the review @go-to-k!
